### ISSUE TITLE
C EventPipe doesn't build incrementally: Changes not propagated to coreclr  

### ISSUE THREAD

**Issue by hoyosjs** [dotnet/runtime#47946]
While working on some eventpipe fundamentals work, @josalem noticed that changes in said source result in non-incremental changes (the translation unit gets compiled, but the changes are not reflected in the resulting coreclr binary).

**Comment by josalem** [comment]
Some more investigation reveals something a little more troublesome: it appears that subsequent builds do not actually compile changed files.

For example:
1. build successfully
2. make a bad change somewhere in src/native/eventpipe (syntax error, whatever)
3. recompile <-- this will succeed and not through compiler errors despite the log saying it saw the file was changed.

**Comment by am11** [comment]
I made a bad change on macOS after successfully running `./build.sh -subset clr+libs -rc release -lc release` once:

```diff
--- a/src/native/eventpipe/ep-buffer-manager.c
+++ b/src/native/eventpipe/ep-buffer-manager.c
@@ -11,7 +11,7 @@
 #include "ep-event-instance.h"
 #include "ep-event-payload.h"
 #include "ep-file.h"
-#include "ep-session.h"
+#include "ep-session.h
 #include "ep-stack-contents.h"
```

the next `./build.sh -subset clr+libs -rc release -lc release` execution results in:

```sh
  [ 88%] Built target crossgen
  In file included from /Users/am11/projects/runtime_pr/src/native/eventpipe/ep.c:10:
  /Users/am11/projects/runtime_pr/src/native/eventpipe/ep-buffer-manager.c:14:10: error: missing terminating '"' character [-Werror,-Winvalid-pp-token]
  #include "ep-session.h
           ^
  /Users/am11/projects/runtime_pr/src/native/eventpipe/ep-buffer-manager.c:14:10: error: expected "FILENAME" or <FILENAME>
  2 errors generated.
  make[2]: *** [vm/eventing/eventpipe/CMakeFiles/eventpipe_objs.dir/Users/am11/projects/runtime_pr/src/native/eventpipe/ep.c.o] Error 1
  make[1]: *** [vm/eventing/eventpipe/CMakeFiles/eventpipe_objs.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
  [ 97%] Built target cee_wks_core
  make: *** [all] Error 2
  Failed to build "CoreCLR component".
/Users/am11/projects/runtime_pr/src/coreclr/runtime.proj(43,5): error MSB3073: The command ""/Users/am11/projects/runtime_pr/src/coreclr/build-runtime.sh" -x64 -release -os OSX" exited with code 2.

Build FAILED.

/Users/am11/projects/runtime_pr/src/coreclr/runtime.proj(43,5): error MSB3073: The command ""/Users/am11/projects/runtime_pr/src/coreclr/build-runtime.sh" -x64 -release -os OSX" exited with code 2.
    0 Warning(s)
    1 Error(s)
```

Perhaps it is a Windows-only issue?

**Comment by josalem** [comment]
> Perhaps it is a Windows-only issue?

That is quite possible, as I didn't encounter this until I was doing some work on Windows.  I haven't tested this behavior on non-Windows yet.

**Comment by hoyosjs** [comment]
This seems to be causing the issues... 

```
#if !defined(EP_INCLUDE_SOURCE_FILES) || defined(EP_FORCE_INCLUDE_SOURCE_FILES)
```

Changing anything within that block on a clean build, or introducing a `#error`, causes build failure. Doing this on an incremental build doesn't. I've only observed this on Windows.

**Comment by josalem** [comment]
Looks to be due to the `{EP|DS}_INCLUDE_SOURCE_FILES` defines.  When those are on, all of the source files are `#include`d into `{ep|ds}.c` and the individual obj files are left empty.  On Windows, it looks like this is somehow causing MSVC to not recompile the `{ep|ds}.c` file on subsequent builds.

**Comment by josalem** [comment]
Confirmed: if you comment out the above define, incremental builds on release work again.

**Comment by am11** [comment]
@lateralusX, just a guess; would it make sense to move this define https://github.com/dotnet/runtime/blob/e29b83934e7ceef9905f82bd6328cbc72db2278b/src/native/eventpipe/ep.c#L6-L7 from ep.c to a header file which is included in .c files using `#if !defined(EP_INCLUDE_SOURCE_FILES) || defined(EP_FORCE_INCLUDE_SOURCE_FILES)` condition? Currently, it seems that the objects (and defines) from ep.c file does not change on rebuild. Alternatively, we can use `DEPENDS` semantics in cmake to make ep.c compile again on rebuild.

**Comment by lateralusX** [comment]
So I believe this is an issue with MSVC and dependency analysis of changed include files. If we have compiler with good LTCG we could disable the include source file "feature" on platforms supporting/using LTGC and have issues detecting changes in include files (Windows seems to be Windows/MSVC only at this moment). I experimented a little locally and believe I have a fix that makes MSVC do what expected.  So there are a number of ways this can be handled. I start with trying a small adjustment to see if that mitigates the issue under MSVC. If it does we can then decide if we should disable the source include feature on some platforms, under some build configs (like debug/checked build) etc.

**Comment by lateralusX** [comment]
Following PR seems to mitigate this issue under MSVC, https://github.com/dotnet/runtime/pull/48050

**Comment by lateralusX** [comment]
I run it on both Mono and CoreClr using MSVC and changes to any EP source file triggered a rebuild of ep-sources.c (broke out into separate source file only doing the include source file feature). It is still possible to locally comment out EP_INCLUDE_SOURCE_FILES on ep-rt-config.h to disable this feature during local development (normally what I do). Passing the define controlling the include of source files in ep-sources.c and ds-sources.c as a pre-processor argument instead of setting it in source file makes MSVC dependency analysis detect changes in included sources and rebuild whats needed.

**Comment by josalem** [comment]
fixed by #48050
